File: pipeline/utils/arxiv.py
"""arXiv ID parsing - Python port of src/lib/arxiv.ts."""

import re
from urllib.parse import urlparse

_ARXIV_ID_PATTERN = re.compile(
    r"^(?:([a-z-]+(?:\.[a-z-]+)?/\d{7})|(\d{4}\.\d{4,5}))(?:v\d+)?(?:\.pdf)?$",
    re.IGNORECASE,
)


def _parse_arxiv_id(value: str) -> str | None:
    m = _ARXIV_ID_PATTERN.match(value.strip())
    if not m:
        return None
    return m.group(1) or m.group(2) or None


def is_arxiv_url(url: str) -> bool:
    try:
        parsed = urlparse(url)
        return bool(re.search(r"(^|\.)arxiv\.org$", parsed.hostname or "", re.IGNORECASE))
    except Exception:
        return False


def extract_arxiv_id(url_or_id: str) -> str | None:
    """Extract the arXiv ID from a URL or raw arXiv ID string."""
    direct = _parse_arxiv_id(url_or_id)
    if direct:
        return direct
    if not is_arxiv_url(url_or_id):
        return None
    try:
        parsed = urlparse(url_or_id)
        pathname = (parsed.path or "").lstrip("/")
        segments = pathname.split("/")
        if len(segments) < 2:
            return None
        kind = segments[0]
        if kind not in ("abs", "pdf"):
            return None
        return _parse_arxiv_id("/".join(segments[1:]))
    except Exception:
        return None


def to_canonical_pdf_url(url_or_id: str) -> str | None:
    """Normalize to standard arXiv PDF URL."""
    arxiv_id = extract_arxiv_id(url_or_id)
    if not arxiv_id:
        return None
    return f"https://arxiv.org/pdf/{arxiv_id}.pdf"

File: src/lib/arxiv.ts
const ARXIV_ID_PATTERN =
  /^(?:([a-z-]+(?:\.[a-z-]+)?\/\d{7})|(\d{4}\.\d{4,5}))(?:v\d+)?(?:\.pdf)?$/i;

function parseArxivId(value: string): string | null {
  const match = value.trim().match(ARXIV_ID_PATTERN);
  if (!match) return null;
  return match[1] ?? match[2] ?? null;
}

export function isArxivUrl(url: string): boolean {
  try {
    const parsed = new URL(url);
    return /(^|\.)arxiv\.org$/i.test(parsed.hostname);
  } catch {
    return false;
  }
}

/**
 * Extract the arXiv ID from a URL or raw arXiv ID.
 * Handles formats like:
 *   https://arxiv.org/abs/2301.12345
 *   https://arxiv.org/pdf/2301.12345
 *   https://arxiv.org/pdf/2301.12345.pdf
 *   https://arxiv.org/pdf/2301.12345v2
 *   https://arxiv.org/pdf/2301.12345v2.pdf
 *   2301.12345
 */
export function extractArxivId(urlOrId: string): string | null {
  const directId = parseArxivId(urlOrId);
  if (directId) return directId;
  if (!isArxivUrl(urlOrId)) return null;

  try {
    const parsed = new URL(urlOrId);
    const pathname = parsed.pathname.replace(/^\/+/, "");
    const segments = pathname.split("/");
    if (segments.length < 2) return null;

    const [kind, ...rest] = segments;
    if (kind !== "abs" && kind !== "pdf") return null;
    return parseArxivId(rest.join("/"));
  } catch {
    return null;
  }
}

export function toCanonicalArxivPdfUrl(urlOrId: string): string | null {
  const arxivId = extractArxivId(urlOrId);
  if (!arxivId) return null;
  return `https://arxiv.org/pdf/${arxivId}.pdf`;
}
